lib/srv/desktop/rdp/rdpclient/README.md

```diff
@@ -85,150 +85,29 @@ hops from one side to the other.
-more or less whatever we want in `unsafe`, and you can do more with mutable references, there's
-a temptation to get a `&mut Client` at the top of `call_into_rust_from_go` like
-
-```rust
-#[no_mangle]
-pub unsafe extern "C" fn call_into_rust_from_go(client_ptr: *mut Client) -> CGOErrCode
-where
-    Client: Send + Sync,
-{
-    // Get immutable reference of Client
-    let client: &mut Client = Box::leak(Box::from_raw(client_ptr));
-    // Do stuff with our immutable reference to Client
-    client.do_mutable_stuff()
-}
-```
-
-Just because you can do something doesn't mean you should, and this is a thing you shouldn't do
-because it breaks all sorts of typical Rust semantics downstream. If you need mutable borrows then
-you'll need to use a `Mutex` and look out for deadlocks.
-
-#### An analogy to pure Rust
-
-Go "owns" the `Client`, and when it calls into Rust with `Client` it essentially is [locking that
-call onto a thread](https://stackoverflow.com/questions/28354141/c-code-and-goroutine-scheduling/28354879#28354879)
-with a `Client` ref (`&Client`) until it returns. A helpful way to reason about this is to imagine how these semantics
-would be instantiated in a pure Rust program.
-
-A naive way to try and do this is [like so](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=91456b6198de6394ecc5b68f1210db70).
-However this doesn't work because
-
-```
-   Compiling playground v0.0.1 (/playground)
-error[E0373]: closure may outlive the current function, but it borrows `client`, which is owned by the current function
-  --> src/main.rs:26:32
-   |
-26 |     let handle = thread::spawn(|| {
-   |                                ^^ may outlive borrowed value `client`
-...
-31 |         println!("{}", client.get_some_field());
-   |                        ------ `client` is borrowed here
-   |
-note: function requires argument type to outlive `'static`
-```
-
-In other words, the compiler can't guarantee that the memory underlying the `&Client` wont be freed before the thread finishes,
-so it disallows this entirely.
-
-The solution for this in Rust is to use [scoped threads](https://doc.rust-lang.org/std/thread/fn.scope.html), which ensure the compiler
-that the threads are all joined by the end of the scope, allowing you to pass references into the threads. So semantically, what we should
-be aiming for is [a program analogous to this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b31ad74edfeb4e2c90551f728fcbcb32).
-
-The key here is that everything at the top level scope of `main` must be implemented by Go; that is to say Go is responsible for ensuring that it drops
-the `Client` at the end of the program, and importantly that it doesn't do so until the threads it spawns (`call_into_rust_from_go`) have returned ("joined").
+Each Go `rdpclient.Client` has a corresponding Rust `client::Client` (`client.rs`). The Go
```

Comment:
This just feels way too complicated. As you read through this and try to keep rack of all the different names and follow links to code it becomes very easy get lost.

Can we bring this up a level of abstraction and talk about the overall model of how things are structured without talking so much about specific parts of code?

Reply:
Try 093784f4b70b2c31e260e8d082de39f070b2cc16